src/jules_pytk/config/inputs.py

```diff
@@ -0,0 +1,117 @@
+    "JulesInput",
+]
+
+def read_ascii(path: str | PathLike) -> numpy.ndarray:
+    with open(path, "r") as file:
+
+        # TODO: comment not being used
+        first_line = file.readline().strip()
+        if first_line[0] in ("#", "!"):  # Treated as comments by JULES
+            comment = first_line
+        else:
+            comment = ""
+
+    data = numpy.loadtxt(path, comments=("#", "!"))
+
+    return data
+
+
+def write_ascii(data: numpy.ndarray, path: str | PathLike, comment="") -> None:
+    numpy.savetxt(path, data, header=comment)
+
+
+def read_netcdf(path: str | PathLike, lazy: bool = True) -> xarray.Dataset:
+    # NOTE: loads entire dataset into memory. Bad idea for big files
+    # TODO: lazy
+    return xarray.load_dataset(path)
+
+
+def write_netcdf(data: xarray.Dataset, path: str | PathLike) -> None:
+    data.to_netcdf(path)
+
+
+class JulesInput:
+    def __init__(self, path_in_namelist: str | PathLike) -> None:
+        path_in_namelist = Path(path_in_namelist)
+
+        match Path(path_in_namelist).suffix:
+            case ".nc" | ".cdf":
+                self._loader = read_netcdf
+                self._dumper = write_netcdf
+            case ".asc" | ".txt" | ".dat":
+                self._loader = read_ascii
+                self._dumper = write_ascii
+            case _:
+                raise ValueError(f"Invalid suffix: {path_in_namelist.suffix}")
+
+        self._path = path_in_namelist
+        self._data = None
+
+    def __eq__(self, other: Any) -> bool:
+        if type(other) != type(self):
+            return False
+        if self.path != other.path:
+            return False
+        if self.data is None:
+            return other.data is None
+        if type(self.data) != type(other.data):
+            return False
+        if isinstance(self.data, numpy.ndarray):
+            # NOTE: consider allclose; these are float arrays
+            return numpy.array_equal(self.data, other.data)
+        if isinstance(self.data, xarray.Dataset):
+            return self.data.identical(other.data)
+
+
+    @property
+    def path(self) -> Path:
+        return self._path
+
+    @property
+    def data(self) -> Any:
+        return self._data
+
+    @data.setter
+    def data(self, new_data) -> None:
+        if self.path.is_absolute():  # and self.path.exists() ?
+            raise Exception("Cannot set data")
+        if self._data is not None:
+            # TODO: Do we want this? If so, clarify with custom exc
+            raise Exception("data is already set. Create a new instance.")
+
+        # TODO: some validation?
+
+        self._data = new_data
+
+    def validate(self, namelists: JulesNamelists) -> bool:
+        # TODO: should this be called in data setter?
+        # NOTE: the idea is to subclass JulesInput and create
+        # validators specific to each type of input file, e.g.
+        # for driving data etc. This seems like a very big task though.
+        return NotImplemented
```

Comment:
The idea here is that one could potentially subclass `JulesInput` to make classes for a specific parameter or parameter group in Jules, which could have custom validation.

For example, the tile fractions input should have some basic validation regarding the number of elements, that elements sum to one etc. Ascii driving data would check for a 1x1 grid, correct number of rows etc, expected variables. So on.

This could turn into quite a large effort.

I'm not against doing it, but need to think more broadly about the best way to validate Jules configurations.